src/sparsemax.jl

```diff
@@ -0,0 +1,90 @@
+"""
+    sparsemax(x; dims = 1)
+
+[Sparsemax](https://arxiv.org/abs/1602.02068) turns input array `x`
+into sparse probability distributions that sum to 1 along the dimensions specified by `dims`.
+
+Similar to softmax, each dimension is considered independent. For a matrix input `x` it will 
+by default (`dims = 1`) treat it as a batch of vectors, with each column independent. Keyword 
+`dims = 2` will instead treat rows independently.
+
+# Examples
+
+```jldoctest
+julia> sparsemax(reshape(1:15, 5,3) ./ 2) # dims=1
+5×3 Matrix{Float64}:
+ 0.0   0.0   0.0
+ 0.0   0.0   0.0
+ 0.0   0.0   0.0
+ 0.25  0.25  0.25
+ 0.75  0.75  0.75
+
+julia> sparsemax(reshape(1:15, 5, 3) ./ 20) # Scale matters
+5×3 Matrix{Float64}:
+ 0.1   0.1   0.1
+ 0.15  0.15  0.15
+ 0.2   0.2   0.2
+ 0.25  0.25  0.25
+ 0.3   0.3   0.3
+
+ julia> sparsemax(reshape(1:15, 5, 3) ./ 20 .+ [-20 0 20])
+ 5×3 Matrix{Float64}:
+  0.1   0.1   0.1
+  0.15  0.15  0.15
+  0.2   0.2   0.2
+  0.25  0.25  0.25
+  0.3   0.3   0.3
+```
+"""
+function sparsemax(x::AbstractArray; dims::Integer=1)
+    z = sort(float(x); dims, rev=true)   # float is usually free, except on integers etc.
+    mask = _sparsemax_mask(z, dims)
+    tausum = sum(z .* mask; dims)  # no longer need z
+    kay = sum(mask; dims)
+    z = _relu.(x  .- (tausum .- 1) ./ kay)
+end
+
+function sparsemax(x::AbstractVector; dims::Integer=1)
+    if dims == 1
+        z = sort(float(x); rev=true)
+    elseif dims == 2
+        z = float(x)
+    end
+    mask = _sparsemax_mask(z, dims)
+    tausum = sum(z .* mask; dims)  # no longer need z
+    kay = sum(mask; dims)
+    z = _relu.(x  .- (tausum .- 1) ./ kay)
+end
+
+
+
+function _sparsemax_mask(z::AbstractArray, dim::Integer)
+    acc = cumsum(z; dims=dim)
+    if dim == 1
+        acc = 1 .+ axes(z,1) .* z .> acc
+    elseif dim == 2
+        acc = 1 .+ axes(z,2)' .* z .> acc
+    else
+        # This isn't type-stable. Writing into `acc` ensures the whole function still is:
+        cnt = reshape(axes(x, dim), ntuple(_->1, dim-1)..., :)
+        acc = 1 .+ cnt .* z .> acc
```

Comment:
Is it the bit mask that adds the instability? Also maybe use `one(T)` where T is inferred from the method signature

Reply:
We get some type instability with the cnt. 

```julia
julia> z = rand(2,3,4);
julia> dim=1; cnt = reshape(axes(z, dim), ntuple(_->1, dim-1)..., :)
Base.OneTo(2)
julia> dim=2; cnt = reshape(axes(z, dim), ntuple(_->1, dim-1)..., :)
1×3 reshape(::Base.OneTo{Int64}, 1, 3) with eltype Int64:
 1  2  3
julia> dim=3; cnt = reshape(axes(z, dim), ntuple(_->1, dim-1)..., :)
1×1×4 reshape(::Base.OneTo{Int64}, 1, 1, 4) with eltype Int64:
[:, :, 1] =
 1

[:, :, 2] =
 2

[:, :, 3] =
 3

[:, :, 4] =
 4
```

Reply:
The mask creation is in a separate function precisely to contain any instability. It writes back into `acc` because this is an easy way to make its return type inferable (and to save allocations). And it avoids calling this reshape in the most common cases, as that made quite a difference to speed.